python/pyiceberg/expressions/__init__.py

```diff
@@ -281,249 +361,313 @@ def __invert__(self) -> BoundIsNull:
+
+    def __eq__(self, other: Any) -> bool:
+        return self.term == other.term and self.literals == other.literals if isinstance(other, SetPredicate) else False
+
+    @property
+    @abstractmethod
+    def as_bound(self) -> Type[BoundSetPredicate[L]]:
+        return BoundSetPredicate[L]
+
+
+class BoundSetPredicate(BoundPredicate[L], ABC):
+    literals: Set[Literal[L]]
 
-@dataclass(frozen=True)
-class BoundSetPredicate(BoundPredicate[T]):
-    literals: set[Literal[T]]
+    def __init__(self, term: BoundTerm[L], literals: Union[Iterable[L], Iterable[Literal[L]]]):
+        # Since we don't know the type of BoundPredicate[L], we have to ignore this one
+        super().__init__(term)  # type: ignore
+        self.literals = _to_literal_set(literals)  # pylint: disable=W0621
 
-    def __invert__(self) -> BoundSetPredicate[T]:
-        """Inverted expression of the SetPredicate"""
-        raise NotImplementedError
+    def __str__(self):
+        # Sort to make it deterministic
+        return f"{str(self.__class__.__name__)}({str(self.term)}, {{{', '.join(sorted([str(literal) for literal in self.literals]))}}})"
 
+    def __repr__(self) -> str:
+        # Sort to make it deterministic
+        return f"{str(self.__class__.__name__)}({repr(self.term)}, {{{', '.join(sorted([repr(literal) for literal in self.literals]))}}})"
 
-@dataclass(frozen=True)
-class BoundIn(BoundSetPredicate[T]):
-    def __new__(cls, term: BoundTerm[T], literals: set[Literal[T]]):  # pylint: disable=W0221
-        count = len(literals)
+    def __eq__(self, other: Any) -> bool:
+        return self.term == other.term and self.literals == other.literals if isinstance(other, BoundSetPredicate) else False
+
+
+class BoundIn(BoundSetPredicate[L]):
+    def __new__(cls, term: BoundTerm[L], literals: Union[Iterable, Iterable[Literal[L]]]):  # pylint: disable=W0221
+        literals_set: Set[Literal[L]] = _to_literal_set(literals)
+        count = len(literals_set)
         if count == 0:
             return AlwaysFalse()
         elif count == 1:
-            return BoundEqualTo(term, next(iter(literals)))
+            return BoundEqualTo(term, next(iter(literals_set)))
         else:
             return super().__new__(cls)
 
-    def __invert__(self) -> BoundNotIn[T]:
+    def __invert__(self) -> BoundNotIn[L]:
         return BoundNotIn(self.term, self.literals)
 
+    def __eq__(self, other: Any) -> bool:
+        return self.term == other.term and self.literals == other.literals if isinstance(other, BoundIn) else False
 
-@dataclass(frozen=True)
-class BoundNotIn(BoundSetPredicate[T]):
-    def __new__(cls, term: BoundTerm[T], literals: set[Literal[T]]):  # pylint: disable=W0221
-        count = len(literals)
+
+class BoundNotIn(BoundSetPredicate[L]):
+    def __new__(  # pylint: disable=W0221
+        cls,
+        term: BoundTerm[L],
+        literals: Union[Iterable[L], Iterable[Literal[L]]],
```

Comment:
Same here, I think we only need the second type.

Reply:
Agreed!